Fix some potential crashes from cantFail/report_fatal_error

benlangmuir opened the pull request:
Remove cantFail and report_fatal_error for CAS Errors that are not programming errors, and propagate the errors instead. Add a new ErroringCAS test helper to inject Error values, and test that simple CachingOnDiskFileSystem and DependencyScanningCASFileSystem operations will not crash and propagate errors correctly.

We probably don't want every possible API using a CAS to need this kind of test, but we can add them to high-level places like these.

Most of the code changes are related to adding TreeSchema::create that returns Expected, so it may be easier to review commit-by-commit.

benlangmuir committed:
[llvm][cas] Add ProxyObjectStore to enable CAS wrappers

Add a ProxyObjectStore that allows wrapping an existing CAS and
forwaring all methods to it, to enable sub-classes that only want to
customize a few operations. This is similar in spiriti to
ProxyFileSystem.

benlangmuir committed:
[llvm][cas] Fix potential crashes in TreeSchema, HierarchicalTreeBuilder, CachingOnDiskFileSystem

Remove cantFail and report_fatal_error for CAS Errors that are not
programming errors, and propogate the errors instead. Add a new
ErroringCAS test helper to inject Error values, and test that simple
CachingOnDiskFileSystem operations that involve all these APIs will not
crash, and that errors are propogated.

We probably don't want every possible API using a CAS to need this kind
of test, but we can add them to high-level places like
CachingOnDiskFileSystem to find failures.

benlangmuir committed:
[clang][cas] [llvm][cas] Fix potential crashes in DependencyScanningCASFileSystem

Avoid report_fatal_error for Errors that could occur in the CAS and
instead propogate them to clang's diagnostic engine. They're still
"fatal" errors, but now they don't bring down the whole process, which
is critical for libclang.

rdar://102869187

benlangmuir commented:
@swift-ci please test

benlangmuir commented on clang/include/clang/Tooling/DependencyScanning/DependencyScanningCASFilesystem.h:
```diff
@@ -72,11 +75,17 @@ class DependencyScanningCASFilesystem : public llvm::cas::ThreadSafeFileSystem {
+  /// Set a \c DiagnosticsEngine to use if there are errors scanning for
+  /// directives. If null, only filesystem errors will be reported via
+  /// the usual \c error_code returns.
+  void setDiagnostics(DiagnosticsEngine *Diags) { this->Diags = Diags; }
```
I don't love this `setDiagnostics` code, but `DependencyScanningCASFilesystem` is shared across multiple scanning TUs whereas the diagnostics are scoped to a single TU.  We could maybe extract the caching layer from the FS to share and create a new FS for each TU so the diagnostics are available at construction time, but that seems like a big refactoring and not really worth it just for this case.

cachemeifyoucan commented on clang/lib/Tooling/DependencyScanning/DependencyScanningCASFilesystem.cpp:
```diff
@@ -140,20 +131,38 @@ void DependencyScanningCASFilesystem::scanForDirectives(
   using namespace llvm;
   using namespace llvm::cas;
 
+  auto reportCASError = [&](unsigned DiagID, Error Err) {
```
What is going to happen if libclang issue a diagnostics? Can build system capture the diagnostics and display it correctly?

Or is it too much to make the change to propagate the errors all the way out and handle it on the top layer (either in clang-scan-deps or build system?)

benlangmuir replied:
We return a `CXDiagnosticSet` from the libclang API when there are errors. There should already be a diagnostic consumer threaded through for this, so I don't think anything extra is needed here.

cachemeifyoucan replied:
If that is the case, then maybe we should enforce the Diagnostics in `reportCASError` otherwise abort? This `consumeError` will setup weird escapes.

benlangmuir replied:
Good idea, it was only a unit test that didn't have a diagnostic engine already.

cachemeifyoucan reviewed (commented):
LGTM in general, except I would prefer handle errors from DepScanCASFS errors outside the FS (either send back through libclang API or Diagnose on a higher level).

benlangmuir commented:
@swift-ci please test

benlangmuir commented:
@swift-ci please test

benlangmuir commented:
@swift-ci please test

benlangmuir commented:
* DiagnosticsEngine now mandatory
* Fixed TreeSchema build error in lld that I missed
* Fixed `return Err` build error that seems to affect older compilers

benlangmuir commented:
@swift-ci please test
